### ISSUE TITLE
Deploying a REST-style API Gateway app fails on 8.4.10

### ISSUE THREAD

**Issue by filmaj** [architect/architect#1089]
**Describe the issue**
Trying to deploy an app with `apigateway` set to `rest` will blow up with the error below. This will also happen even if you don't have `apigateway rest` in your `app.arc` file, but have previously deployed the app back when the REST-style API Gateway was the default APIG configuration (as `deploy`'s `src/sam/compat` module [will check CloudFormation for the kind of API Gateway configuration the app already uses](https://github.com/architect/deploy/blob/master/src/sam/compat/index.js#L56-L70)).

```
➜ arc deploy
         App ⌁ vintappd
      Region ⌁ ca-central-1
     Profile ⌁ default
     Version ⌁ Architect 8.4.10
         cwd ⌁ /Users/filmaj/src/vintappd

⚬ Deploy Creating new private deployment bucket: vintappd-cfn-deployments-25c13
⚬ Deploy Initializing deployment
  | Stack ... VintappdStaging
  | Bucket .. vintappd-cfn-deployments-25c13
⚬ Hydrate Hydrating dependencies in 7 paths
✓ Hydrate Hydrated src/http/get-vp-000vpCode-link/
  | npm ci --production: added 111 packages in 1.406s
✓ Hydrate Hydrated src/http/get-vp-000vpCode/
  | npm ci --production: added 113 packages in 1.114s
✓ Hydrate Hydrated src/http/post-vp-000vpCode-refresh/
  | npm ci --production: added 112 packages in 1.284s
✓ Hydrate Hydrated src/http/post-vp-000vpCode/
  | npm ci --production: added 113 packages in 1.191s
✓ Hydrate Hydrated src/http/post-vp-autolink-000vpCode/
  | npm ci --production: added 110 packages in 1.203s
✓ Hydrate Hydrated src/http/post-vp-getBatch/
  | npm ci --production: added 101 packages in 0.704s
✓ Hydrate Hydrated src/shared/
  | npm ci --production: added 109 packages in 1.442s
⚬ Hydrate Hydrating app with shared files
✓ Hydrate Hydrated app with shared: src/shared
✓ Hydrate Successfully hydrated dependencies
deploy failed! Cannot read property 'Properties' of undefined
TypeError: Cannot read property 'Properties' of undefined
    at /Users/filmaj/src/vintappd/node_modules/@architect/deploy/src/sam/macros/_legacy-api/index.js:61:48
    at Array.forEach (<anonymous>)
    at legacyAPI (/Users/filmaj/src/vintappd/node_modules/@architect/deploy/src/sam/macros/_legacy-api/index.js:39:10)
    at reducer (/Users/filmaj/src/vintappd/node_modules/@architect/deploy/src/sam/macros/index.js:71:20)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async reducer (/Users/filmaj/src/vintappd/node_modules/@architect/deploy/src/sam/macros/index.js:70:28)
    at async reducer (/Users/filmaj/src/vintappd/node_modules/@architect/deploy/src/sam/macros/index.js:70:28)
    at async reducer (/Users/filmaj/src/vintappd/node_modules/@architect/deploy/src/sam/macros/index.js:70:28)
    at async reducer (/Users/filmaj/src/vintappd/node_modules/@architect/deploy/src/sam/macros/index.js:70:28)
```

```
➜ cat .arc
@app
vintappd

@http
post /vp/getBatch
post /vp/:vpCode
post /vp/autolink/:vpCode
post /vp/:vpCode/refresh
get /vp/:vpCode
get /vp/:vpCode/link

@tables
beers
  vpCode *String

@aws
runtime nodejs12.x
region ca-central-1
apigateway rest
```

**Steps to reproduce**
1. Create an arc app with the `app.arc` file posted above
2. Run `arc deploy --dry-run --no-hydrate`

**Expected behavior**
No 💥

**Comment by filmaj** [comment]
The offending line is [here](https://github.com/architect/deploy/blob/master/src/sam/macros/_legacy-api/index.js#L60):

```
cloudformation.Resources[routeLambdaOld] = cloudformation.Resources[routeLambdaNew]
cloudformation.Resources[routeLambdaOld].Properties.Events[routeEventOld] = {
```

That second line referencing `Properties` is where the error happens.

For the `app.arc` file I posted above, I added a `console.log` to see what the values for the Resource names were at the time of the explosion. Here's what I found:

```
⚬ Deploy [dry-run] Initializing deployment
  | Stack ... VintappdStaging
  | Bucket .. N/A (dry-run)
post /vp/getBatch PostVpGetBatch PostVpGetBatchHTTPLambda PostVpGetBatchEvent PostVpGetBatchHTTPEvent
post /vp/{vpCode} PostVpVpCode PostVpVpCodeHTTPLambda PostVpVpCodeEvent PostVpVpCodeHTTPEvent
post /vp/autolink/{vpCode} PostVpAutolinkVpCode PostVpAutolinkVpCodeHTTPLambda PostVpAutolinkVpCodeEvent PostVpAutolinkVpCodeHTTPEvent
post /vp/{vpCode}/refresh PostVpVpCodeRefresh PostVpVpCodeRefreshHTTPLambda PostVpVpCodeRefreshEvent PostVpVpCodeRefreshHTTPEvent
get /vp/{vpCode} GetVpVpCode GetVpVpCodeHTTPLambda GetVpVpCodeEvent GetVpVpCodeHTTPEvent
get /vp/{vpCode}/link GetVpVpCodeLink GetVpVpCodeLinkHTTPLambda GetVpVpCodeLinkEvent GetVpVpCodeLinkHTTPEvent
get / GetIndex GetIndexHTTPLambda GetIndexEvent GetIndexHTTPEvent
deploy failed! Cannot read property 'Properties' of undefined
TypeError: Cannot read property 'Properties' of undefined
    at /Users/filmaj/src/vintappd/node_modules/@architect/deploy/src/sam/macros/_legacy-api/index.js:62:48
```

So for the 6 user-defined HTTP routes, this `legacy api` build-in macro executes flawlessly. However, it fails at the `get /` route. This route is not explicitly defined by the user, however, the legacy-api macro [adds it at the beginning of execution if it is not present](https://github.com/architect/deploy/blob/master/src/sam/macros/_legacy-api/index.js#L15-L20):

```
    // Force add GetIndex if not defined
    let findGetIndex = tuple => tuple[0].toLowerCase() === 'get' && tuple[1] === '/'
    let hasGetIndex = http.some(findGetIndex) // we reuse this below for default proxy code
    if (!hasGetIndex) {
      http.push([ 'get', '/' ])
    }
```

So in order to reproduce this bug, you must define a REST API Gateway app (or have deployed one in the past) and not have a `get /` route defined explicitly. Since legacy-api adds the route to the arc file at runtime, but doesn't add it to the CloudFormation template, this is where the discrepancy exists and leads to this bug.

**Comment by filmaj** [comment]
If I additionally log out the keys present in `cloudformation.Resources` right before this exception happens, here is what I see:

```
[
  'GetCatchallHTTPLambda',
  'BeersTable',
  'StaticBucket',
  'Role',
  'BeersParam',
  'StaticBucketParam',
  'StaticFingerprintParam',
  'ParameterStorePolicy',
  'Vintappd',
  'PostVpGetBatch',
  'PostVpVpCode',
  'PostVpAutolinkVpCode',
  'PostVpVpCodeRefresh',
  'GetVpVpCode',
  'GetVpVpCodeLink'
]
```

So there is no `GetIndex` route, but there is a `GetCatchallHTTPLambda` one (which seems to point to the built-in arc `asap` code?). Is that perhaps where the confusion is?

**Comment by ryanblock** [comment]
Almost certainly, iirc there was a check in there for this but maybe I lost or broke it at some point. Anyhow, ASAP for REST should be mounted to get /, not get /*. I'll take a look today (I hope!).

**Comment by ryanblock** [comment]
Ok, well that sat way longer than expected, thanks for your patience everyone! Fix is now live in 8.5.2 RC (`npm i @architect/architect@RC`). Should be totally safe, but out of an abundance of caution I'd suggest testing that in a staging env (or a test project) to start.

**Comment by filmaj** [comment]
Thanks for getting the fix for this out @ryanblock ! It should be available in arc 8.5.2.
